### PR TITLE
Add support for command-specific configuration sections

### PR DESCRIPTION
**Pull request by chrysle**
Addresses https://github.com/jazzband/pip-tools/pull/1933#issuecomment-1653068755

This introduces configuration sections specific to `pip-compile` and `pip-sync`.

I'd like to refactor the `make_config_file()` pytest fixture in a follow-up PR, to introduce tests for this feature using multiple configuration sections.

##### Contributor checklist

- [x] Included tests for the changes.
- [x] PR title is short, clear, and ready to be included in the user-facing changelog.

##### Maintainer checklist

- [x] Verified one of these labels is present: `backwards incompatible`, `feature`, `enhancement`, `deprecation`, `bug`, `dependency`, `docs` or `skip-changelog` as they determine changelog listing.
- [ ] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

**Comment by webknjaz** [comment]
Could you adjust the PR title to match the description? Currently it reads as if you're adding features but the text suggests that it's tests.

**Comment by chrysle** [comment]
Thanks, it's actually **this** pull request I'd like to add more tests for (codecov is failing), but that needs some refactoring and doesn't really fit in. I've adjusted the description.

**Comment by webknjaz** [comment]
Ah, though I still don't understand what's tool-specific configuration. That title is just to generic to mean anything to me. It's quite unspecific wrt what it refers to.

**Comment by chrysle** [comment]
>  We already have pip-tools specific config, after all. That's the tool, right?

Yeah, but that is the _global_ configuration (for both `pip-compile` and `pip-sync`). I added the possibility to specify configuration values _only for_ those commands. I added some docs to the README.

Is the title better now?

**Comment by webknjaz** [comment]
> Is the title better now?

Much better, thanks! Titles shows up in email notifications among other places and tend to be a factor in deciding whether to even look inside. This is especially important when inbox is overflowing and people need to prioritize + figure out which notifications to discard. An accurate title helps draw attention of people who might be affected by the patch.

File: README.md
Comment (webknjaz) [inline README.md:293]: Let's avoid saying tool-specific since the document doesn't record what is a "tool" in different contexts and people may be getting different incorrect ideas.
"command" is probably fine. Could also be a sub-command, I suppose. Let's see if others have better ideas, though...

File: README.md
Comment (webknjaz) [inline README.md:324]: Let's be more specific than "overrides". When I read this, I thought that the entire global section is discarded. We need to choose wording to highlight that the sections are merged and individual specific settings supersede / take precedence over the global ones.

File: piptools/utils.py
Comment (webknjaz) [inline piptools/utils.py:573]: I don't think it's a good idea to return a tuple. If it is really necessary to have all these bits of data, wrap them into an object (dataclass?) and return that.

Though, I'd argue it should return a single config instead. Maybe, we need to pass an argument telling it for which command that config is to be extracted..

File: piptools/utils.py
Comment (webknjaz) [inline piptools/utils.py:717]: Traditionally, everything concerning a project, should be a part of the same namespaces. We can use subsections for this:

```suggestion
    # `[tool.pip-tools.pip-compile]` or `[tool.pip-tools.pip-sync]`
```

File: tests/test_cli_sync.py
Comment (webknjaz) [inline tests/test_cli_sync.py:455]: Don't use `mock` directly. In pytest, use the `mocker` fixture or `monkeypatch` if possible.

**Review by webknjaz — COMMENTED** [review]
Initial review. I foresee the need to refactor things here and avoid complicating things.

File: tests/test_cli_sync.py
Comment (chrysle) [reply tests/test_cli_sync.py:455]: I've adapted the test style to the rest of the file.

File: piptools/utils.py
Comment (chrysle) [reply piptools/utils.py:717]: Though it's pretty tedious for the user to type this..

File: piptools/utils.py
Comment (chrysle) [reply piptools/utils.py:573]: Yes, I can move the test below to `parse_config_file()`.

File: README.md
Comment (chrysle) [reply README.md:324]: How about `conflicting settings in the global section get overriden by the local ones`?

File: piptools/utils.py
Comment (theryanwalker) [reply piptools/utils.py:717]: "Namespaces are one honking great idea". Wouldn't it be more appropriate to use `[tool.pip-tools.compile]`, similar to `python -m piptools.compile`?

File: piptools/utils.py
Comment (theryanwalker) [reply piptools/utils.py:573]:
```suggestion
    config = Config.parse_file(config_file, ctx)
```

Given that you're in the process of refactoring this, how about introducing a `Config` class in `piptools/config.py` and moving the related logic there? This is an example of what the Config might look like:

```python
@dataclass
class Config:
    pip_tools: dict[str, Any]
    pip_compile: dict[str, Any]
    pip_sync: dict[str, Any]
    
    @classmethod
    def parse_file(filename):
        ...

    def validate(self):
        ...
        
    def normalize_keys(self):
        ...
    
    def invert_negative_bool_options(self, ctx):
        ...
        
    def get_default_map() -> dict[str, Any]:
        ...
     
    def assign_to_cli(ctx):
        ...    
```

Comment (chrysle) [reply piptools/utils.py:573]: @atugushev What do you think, should we dedicate an own file to that?

Comment (atugushev) [reply piptools/utils.py:573]: Sounds great 👍🏻

File: piptools/utils.py
Comment (webknjaz) [reply piptools/utils.py:717]: @theryanwalker fair. That's a better suggestion.

```suggestion
    # `[tool.pip-tools.compile]` or `[tool.pip-tools.sync]`
```

File: README.md
Comment (webknjaz) [reply README.md:324]: "The local settings take preference over the global ones of the same name, whenever both are declared"?

File: piptools/utils.py
Comment (webknjaz) [reply piptools/utils.py:573]: @theryanwalker I like it, though I'd probably go for a different construction/classmethod name. How about `Config.from_file_path()`?

File: piptools/utils.py
Comment (webknjaz) [inline piptools/utils.py:579]: Is it still `"pip-compile"` when invoked using `python -m piptools compile`?

File: piptools/utils.py
Comment (webknjaz) [inline piptools/utils.py:587]: Let's dedent this a bit:
```suggestion
    for config in filter(bool, configs):
        _validate_config(ctx, config)
        _assign_config_to_cli_context(ctx, config)
```

**Comment by webknjaz** [comment]
@chrysle looks like this also needs to be rebased due to conflicts in tests.

File: tests/test_cli_sync.py
Comment (webknjaz) [reply tests/test_cli_sync.py:455]: Okay, then this is out of the scope of this PR. Marking the thread as resolved.

File: piptools/utils.py
Comment (chrysle) [reply piptools/utils.py]: Yes, verified.

File: README.md
Comment (chrysle) [reply README.md]: Reworded to "command".

File: piptools/utils.py
Comment (webknjaz) [reply piptools/utils.py:682]: @chrysle looks like you forgot to apply the suggested change to the comment line, so I'm reopening this.

File: piptools/utils.py
Comment (webknjaz) [inline piptools/utils.py:689]: No need to double-lookup the dicts. The entire section is already in the var:
```suggestion
    pipcompile_config: dict[str, Any] = piptools_config.get("compile", {})
    pipsync_config: dict[str, Any] = piptools_config.get("sync", {})
```

File: piptools/utils.py
Comment (webknjaz) [inline piptools/utils.py:695]: I would think you'd want to remove the other one…

File: piptools/utils.py
Comment (webknjaz) [inline piptools/utils.py:700]: This can probably be simplified further:
```suggestion
    cmd_to_section_map = {
        "pip-compile": "compile",
        "pip-sync": "sync",
    }
    config_section_name = cmd_to_section_map[click_context.command.name]
    piptools_config.update(piptools_config.pop(config_section_name, {}))
    piptools_config.pop("compile", {})
    piptools_config.pop("sync", {})
```
The above assumes we don't want those subsections in the object we'll be passing around...

Comment (webknjaz) [reply piptools/utils.py:700]: The map isn't really needed, though:
```suggestion
    config_section_name = click_context.command.name.removeprefix("pip-")
    piptools_config.update(piptools_config.pop(config_section_name, {}))
    piptools_config.pop("compile", {})
    piptools_config.pop("sync", {})
```

Comment (webknjaz) [reply piptools/utils.py:700]: This might be easier to grasp:
```suggestion
    config_section_name = click_context.command.name.removeprefix("pip-")
    piptools_config = {
        **piptools_config,
        "compile": {},
        "sync": {},
        **piptools_config.pop(config_section_name, {}),
    }

```

File: piptools/utils.py
Comment (webknjaz) [inline piptools/utils.py:702]: No need to rename the variable really.

File: tests/conftest.py
Comment (webknjaz) [inline tests/conftest.py:519]: How about
```suggestion
        nested_config = {pyproject_param: new_default}
        if subsection:
            nested_config = {subsection: nested_config}
        config_to_dump = {"tool": {section: nested_config}}
```

File: piptools/utils.py
Comment (webknjaz) [inline piptools/utils.py:687]: I think this
```suggestion
    config_section_name = click_context.command.name[len("pip-"):]
```
would be a more accurate representation of what `str.removeprefix()` does.

File: tests/test_cli_compile.py
Comment (webknjaz) [inline tests/test_cli_compile.py:3573]: Is this some leftover debug output?

**Comment by chrysle** [comment]
Thank you for your suggestions and sorry for the long delay.
My code style is simply terrible... My new year's resolution definitely includes paying more attention in computer science class.

File: piptools/utils.py
Comment (webknjaz) [inline piptools/utils.py:686]: The comment is disconnected from the line it refers to
```suggestion
    assert click_context.command.name is not None
    # TODO: Replace with `str.removeprefix()` once dropped 3.8
```

**Review by webknjaz — APPROVED** [review]

**Comment by webknjaz** [comment]
@chrysle hey, no problem! This is open source and nobody _expects_ free labor to be a top priority of every contributor.

Also, the code style is something that evolves and is subjective most of the time. People are only expected to unconditionally adhere what automatic checks say. Otherwise, code reviews is a multi-party discussion that exists to agree on things and receive insights from other people that aren't related to code style.

At times, it may feel like some comments are about code style at glance, but usually, there's more important underlying issues that motivate people to make suggestions. It's mostly a perception thing. I know that my comments are often seen as style nitpicks while I write them with maintainability in mind, informed by architectural considerations, usually.

Here's a few links that I collected about code reviews, by the way: https://gist.github.com/webknjaz/2aa9c7a43b51c1385260ff87e0cbb9e3.

P.S. The `linkcheck` job is failing because `jazzband.co` has a TLS certificate that expired on Dec 30 last year and wasn't auto-renewed for some reason. A tagged @jezdez on matrix so he could take a look.
This blocks merging the PR, but once fixed, restarting that one failing job should unblock us.

**Comment by chrysle** [comment]
Thank you for your encouraging words and the time you invest into reviewing my code!
I'll also still have to solve the mystery how people get the right ideas for founding successful open source projects ;-)

**Pull request merged**
